File: packages/roaster/src/roaster/review_definition.py
"""Parsing helpers for markdown-defined reviewers."""

from __future__ import annotations

from typing import Any

import yaml

from roaster.models import ReviewDefinition

_FRONTMATTER_FENCE = "---"

_ALLOWED_FRONTMATTER_KEYS: frozenset[str] = frozenset(
    {"ci", "description", "default_model", "when_changed"}
)


def parse_review_definition(source: str, *, name: str) -> ReviewDefinition:
    """Parse a review definition written as YAML frontmatter + markdown body.

    The reviewer `name` is supplied by the caller (typically the filename
    without its extension) rather than declared in the frontmatter.
    """
    if not name.strip():
        raise ValueError("Review definition `name` must be a non-empty string.")

    frontmatter_text, body = _split_frontmatter(source)

    try:
        parsed_frontmatter = yaml.safe_load(frontmatter_text)
    except yaml.YAMLError as exc:
        raise ValueError(f"Review definition frontmatter is not valid YAML: {exc}") from exc

    if parsed_frontmatter is None:
        raise ValueError("Review definition frontmatter is empty.")
    if not isinstance(parsed_frontmatter, dict):
        raise ValueError("Review definition frontmatter must be a YAML mapping.")

    _reject_unknown_keys(parsed_frontmatter)

    description = _require_string(parsed_frontmatter, "description")
    ci = _require_ci_bool(frontmatter_text, parsed_frontmatter)

    default_model_value = parsed_frontmatter.get("default_model")
    if default_model_value is None:
        default_model: str | None = None
    elif isinstance(default_model_value, str) and default_model_value.strip():
        default_model = default_model_value.strip()
    else:
        raise ValueError("Review definition field `default_model` must be a non-empty string.")

    when_changed = _optional_string_tuple(parsed_frontmatter, "when_changed")

    instructions = body.strip()
    if not instructions:
        raise ValueError("Review definition body (instructions) must not be empty.")

    return ReviewDefinition(
        name=name.strip(),
        description=description,
        instructions=instructions,
        default_model=default_model,
        ci=ci,
        when_changed=when_changed,
    )


def _reject_unknown_keys(frontmatter: dict[str, Any]) -> None:
    unknown = sorted(key for key in frontmatter if key not in _ALLOWED_FRONTMATTER_KEYS)
    if not unknown:
        return
    allowed = ", ".join(sorted(_ALLOWED_FRONTMATTER_KEYS))
    unknown_list = ", ".join(f"`{key}`" for key in unknown)
    raise ValueError(
        f"Review definition frontmatter contains unknown field(s): {unknown_list}. "
        f"Allowed fields: {allowed}."
    )


def _split_frontmatter(source: str) -> tuple[str, str]:
    lines = source.splitlines()
    first_content_index = next((index for index, line in enumerate(lines) if line.strip()), None)
    if first_content_index is None:
        raise ValueError("Review definition is empty.")

    if lines[first_content_index].strip() != _FRONTMATTER_FENCE:
        raise ValueError("Review definition must begin with a `---` frontmatter fence.")

    closing_index: int | None = None
    for index in range(first_content_index + 1, len(lines)):
        if lines[index].strip() == _FRONTMATTER_FENCE:
            closing_index = index
            break
    if closing_index is None:
        raise ValueError("Review definition frontmatter is missing a closing `---` fence.")

    frontmatter_text = "\n".join(lines[first_content_index + 1 : closing_index])
    body = "\n".join(lines[closing_index + 1 :])
    return frontmatter_text, body


def _require_string(frontmatter: dict[str, Any], field: str) -> str:
    if field not in frontmatter:
        raise ValueError(f"Review definition frontmatter is missing required field `{field}`.")
    value = frontmatter[field]
    if not isinstance(value, str) or not value.strip():
        raise ValueError(f"Review definition field `{field}` must be a non-empty string.")
    return value.strip()


def _require_ci_bool(frontmatter_text: str, frontmatter: dict[str, Any]) -> bool:
    field = "ci"
    if field not in frontmatter:
        raise ValueError(f"Review definition frontmatter is missing required field `{field}`.")

    ci_literal = _top_level_ci_literal(frontmatter_text)
    if ci_literal not in {"true", "false"} or not isinstance(frontmatter[field], bool):
        raise ValueError("Review definition field `ci` must be literal true or false.")
    return frontmatter[field]


def _top_level_ci_literal(frontmatter_text: str) -> str | None:
    ci_lines: list[str] = []
    for line in frontmatter_text.splitlines():
        if line[:1].isspace():
            continue
        stripped = line.strip()
        if stripped.startswith("ci:"):
            ci_lines.append(stripped)

    if len(ci_lines) != 1:
        return None

    value = ci_lines[0].partition(":")[2].strip()
    if "#" in value:
        value = value.split("#", 1)[0].strip()
    return value


def _optional_string_tuple(frontmatter: dict[str, Any], field: str) -> tuple[str, ...]:
    if field not in frontmatter:
        return ()

    value = frontmatter[field]
    if not isinstance(value, list):
        raise ValueError(f"Review definition field `{field}` must be a list of non-empty strings.")
    if not value:
        raise ValueError(f"Review definition field `{field}` must contain at least one pattern.")

    values: list[str] = []
    for index, item in enumerate(value):
        if not isinstance(item, str) or not item.strip():
            raise ValueError(
                f"Review definition field `{field}` item {index} must be a non-empty string."
            )
        values.append(item.strip())

    return tuple(values)

File: ts/packages/roaster/src/review-definition.ts
import { splitMarkdownFrontmatter } from "@asdl/core/markdown-frontmatter";
import { formatErrorMessage, formatZodError, isRecord } from "@asdl/core/primitives";
import { parse } from "yaml";

import {
	reviewDefinitionSchema,
	type ReviewApplicability,
	type ReviewDefinition,
} from "./models.ts";

const ALLOWED_FRONTMATTER_KEYS = ["applies_to", "default_model", "description"] as const;
const ALLOWED_APPLIES_TO_KEYS = ["exclude", "include"] as const;

export type ReviewDefinitionParseResult =
	| { readonly type: "ok"; readonly definition: ReviewDefinition }
	| { readonly type: "error"; readonly error: ReviewDefinitionParseError };

export interface ReviewDefinitionParseError {
	readonly code: ReviewDefinitionParseErrorCode;
	readonly message: string;
}

export type ReviewDefinitionParseErrorCode =
	| "empty_source"
	| "missing_open_fence"
	| "missing_close_fence"
	| "invalid_yaml"
	| "invalid_frontmatter"
	| "unknown_frontmatter_key"
	| "invalid_name"
	| "invalid_description"
	| "invalid_default_model"
	| "invalid_instructions"
	| "invalid_applicability";

interface ParseReviewDefinitionOptions {
	readonly name: string;
}

export function parseReviewDefinition(
	source: string,
	options: ParseReviewDefinitionOptions,
): ReviewDefinitionParseResult {
	const name = options.name.trim();
	if (name === "") {
		return failure("invalid_name", "Review definition `name` must be a non-empty string.");
	}

	const split = splitFrontmatter(source);
	if (split.type === "error") return split;

	let parsedFrontmatter: unknown;
	try {
		parsedFrontmatter = parse(split.frontmatterText);
	} catch (error) {
		return failure(
			"invalid_yaml",
			`Review definition frontmatter is not valid YAML: ${formatErrorMessage(error)}`,
		);
	}

	if (parsedFrontmatter === null || parsedFrontmatter === undefined) {
		return failure("invalid_frontmatter", "Review definition frontmatter is empty.");
	}
	if (!isRecord(parsedFrontmatter)) {
		return failure("invalid_frontmatter", "Review definition frontmatter must be a YAML mapping.");
	}

	const unknownKeys = sortedUnknownKeys(parsedFrontmatter, ALLOWED_FRONTMATTER_KEYS);
	if (unknownKeys.length > 0) {
		const unknownList = unknownKeys.map((key) => `\`${key}\``).join(", ");
		const allowed = [...ALLOWED_FRONTMATTER_KEYS].sort().join(", ");
		return failure(
			"unknown_frontmatter_key",
			`Review definition frontmatter contains unknown field(s): ${unknownList}. Allowed fields: ${allowed}.`,
		);
	}

	const description = requireStringField(parsedFrontmatter, "description");
	if (description.type === "error") return description;

	const defaultModel = parseDefaultModel(parsedFrontmatter);
	if (defaultModel.type === "error") return defaultModel;

	const applicability = parseApplicability(parsedFrontmatter);
	if (applicability.type === "error") return applicability;

	const instructions = split.body.trim();
	if (instructions === "") {
		return failure(
			"invalid_instructions",
			"Review definition body (instructions) must not be empty.",
		);
	}

	const candidate = {
		name,
		description: description.value,
		instructions,
		defaultModel: defaultModel.value,
		applicability: applicability.value,
	};
	const parsedDefinition = reviewDefinitionSchema.safeParse(candidate);
	if (!parsedDefinition.success) {
		throw new Error(
			`Review definition parser produced a value that does not match reviewDefinitionSchema: ${formatZodError(parsedDefinition.error)}`,
		);
	}
	return { type: "ok", definition: parsedDefinition.data };
}

type FrontmatterSplitResult =
	| { readonly type: "ok"; readonly frontmatterText: string; readonly body: string }
	| { readonly type: "error"; readonly error: ReviewDefinitionParseError };

function splitFrontmatter(source: string): FrontmatterSplitResult {
	if (source.trim() === "") return failure("empty_source", "Review definition is empty.");
	const split = splitMarkdownFrontmatter(source);
	if (split.type === "not_found")
		return failure(
			"missing_open_fence",
			"Review definition must begin with a `---` frontmatter fence.",
		);
	if (split.type === "missing_closing_fence")
		return failure(
			"missing_close_fence",
			"Review definition frontmatter is missing a closing `---` fence.",
		);
	return {
		type: "ok",
		frontmatterText: split.block.frontmatterText.replace(/\r\n?/gu, "\n"),
		body: split.block.body.replace(/\r\n?/gu, "\n"),
	};
}

type StringFieldResult =
	| { readonly type: "ok"; readonly value: string }
	| { readonly type: "error"; readonly error: ReviewDefinitionParseError };

function requireStringField(
	frontmatter: Readonly<Record<string, unknown>>,
	field: string,
): StringFieldResult {
	if (!(field in frontmatter)) {
		return failure(
			"invalid_description",
			`Review definition frontmatter is missing required field \`${field}\`.`,
		);
	}
	const value = frontmatter[field];
	if (typeof value !== "string" || value.trim() === "") {
		return failure(
			"invalid_description",
			`Review definition field \`${field}\` must be a non-empty string.`,
		);
	}
	return { type: "ok", value: value.trim() };
}

type DefaultModelResult =
	| { readonly type: "ok"; readonly value: string | null }
	| { readonly type: "error"; readonly error: ReviewDefinitionParseError };

function parseDefaultModel(frontmatter: Readonly<Record<string, unknown>>): DefaultModelResult {
	if (!("default_model" in frontmatter)) return { type: "ok", value: null };
	const value = frontmatter.default_model;
	if (typeof value !== "string" || value.trim() === "") {
		return failure(
			"invalid_default_model",
			"Review definition field `default_model` must be a non-empty string.",
		);
	}
	return { type: "ok", value: value.trim() };
}

type ApplicabilityResult =
	| { readonly type: "ok"; readonly value: ReviewApplicability }
	| { readonly type: "error"; readonly error: ReviewDefinitionParseError };

function parseApplicability(frontmatter: Readonly<Record<string, unknown>>): ApplicabilityResult {
	if (!("applies_to" in frontmatter)) return { type: "ok", value: { include: [], exclude: [] } };
	const value = frontmatter.applies_to;
	if (!isRecord(value)) {
		return failure(
			"invalid_applicability",
			"Review definition field `applies_to` must be a YAML mapping.",
		);
	}

	const unknownKeys = sortedUnknownKeys(value, ALLOWED_APPLIES_TO_KEYS);
	if (unknownKeys.length > 0) {
		const unknownList = unknownKeys.map((key) => `\`${key}\``).join(", ");
		return failure(
			"invalid_applicability",
			`Review definition field \`applies_to\` contains unknown field(s): ${unknownList}.`,
		);
	}

	const include = requirePatternList(value, { field: "include", shouldAllowEmpty: false });
	if (include.type === "error") return include;
	const exclude =
		"exclude" in value
			? requirePatternList(value, { field: "exclude", shouldAllowEmpty: true })
			: { type: "ok" as const, value: [] };
	if (exclude.type === "error") return exclude;
	return { type: "ok", value: { include: include.value, exclude: exclude.value } };
}

interface RequirePatternListOptions {
	readonly field: "include" | "exclude";
	readonly shouldAllowEmpty: boolean;
}

type PatternListResult =
	| { readonly type: "ok"; readonly value: string[] }
	| { readonly type: "error"; readonly error: ReviewDefinitionParseError };

function requirePatternList(
	appliesTo: Readonly<Record<string, unknown>>,
	options: RequirePatternListOptions,
): PatternListResult {
	if (!(options.field in appliesTo)) {
		return failure(
			"invalid_applicability",
			`Review definition field \`applies_to.${options.field}\` is required.`,
		);
	}

	const value = appliesTo[options.field];
	if (!Array.isArray(value)) {
		return failure(
			"invalid_applicability",
			`Review definition field \`applies_to.${options.field}\` must be a list of strings.`,
		);
	}
	if (value.length === 0 && !options.shouldAllowEmpty) {
		return failure(
			"invalid_applicability",
			`Review definition field \`applies_to.${options.field}\` must not be empty.`,
		);
	}

	const patterns: string[] = [];
	for (const pattern of value) {
		const result = validateApplicabilityPattern(pattern, options.field);
		if (result.type === "error") return result;
		patterns.push(result.value);
	}
	return { type: "ok", value: patterns };
}

function validateApplicabilityPattern(
	pattern: unknown,
	field: "include" | "exclude",
): StringFieldResult {
	if (typeof pattern !== "string" || pattern.trim() === "") {
		return failure(
			"invalid_applicability",
			`Review definition field \`applies_to.${field}\` must contain non-empty strings.`,
		);
	}

	const normalized = pattern.trim().replaceAll("\\", "/");
	if (normalized.startsWith(":(")) {
		return failure(
			"invalid_applicability",
			"Review definition applicability patterns must be globs, not git pathspecs.",
		);
	}
	if (normalized.startsWith("/")) {
		return failure(
			"invalid_applicability",
			"Review definition applicability patterns must be repo-relative.",
		);
	}
	if (normalized.split("/").includes("..")) {
		return failure(
			"invalid_applicability",
			"Review definition applicability patterns must not contain `..` segments.",
		);
	}
	return { type: "ok", value: normalized };
}

function failure(
	code: ReviewDefinitionParseErrorCode,
	message: string,
): { readonly type: "error"; readonly error: ReviewDefinitionParseError } {
	return { type: "error", error: { code, message } };
}

function sortedUnknownKeys(
	record: Readonly<Record<string, unknown>>,
	allowedKeys: readonly string[],
): string[] {
	return Object.keys(record)
		.filter((key) => !allowedKeys.includes(key))
		.sort();
}
